File: apps/web/src/lib/telemetry.ts
export type FlowOutcome = "success" | "failure" | "abandoned";

type OperationalEvent = {
  name: string;
  dimension?: string;
  outcome?: string;
  method?: string;
  status?: number;
  durationMs?: number;
};

const API_BASE = import.meta.env.VITE_API_BASE ?? "/api";
const TELEMETRY_ENDPOINT = `${API_BASE.replace(/\/+$/, "")}/telemetry`;
const FLUSH_DELAY_MS = 1_500;
const MAX_BATCH_SIZE = 25;
const SUCCESS_EVENTS = new Set([
  "success_dashboard_opened",
  "success_account_connected",
  "success_pro_trial_activated",
  "success_credits_purchased",
  "success_support_request_sent",
  "success_watchlist_changed",
  "success_portfolio_changed",
  "success_dca_plan_changed",
  "success_ai_analysis_returned",
  "success_strategy_returned",
  "success_replay_started",
  "success_onboarding_completed",
]);
const FLOWS = new Set([
  "hunt_signals",
  "buy_timing",
  "next_10",
  "technical_analysis",
  "intraday_analysis",
  "strategy_analysis",
  "analyst_report",
  "ai_replay",
]);
const PAGE_BY_PATH: Record<string, string> = {
  "/": "home",
  "/terms": "terms",
  "/privacy": "privacy",
  "/refunds": "refunds",
  "/support": "support",
  "/daily-brief": "daily_brief",
  "/live-trade": "live_trade",
  "/scanner": "scanner",
  "/discover": "scanner",
  "/hunt-ai": "hunt_ai",
  "/deep-ai": "hunt_ai",
  "/day-trader": "hunt_ai",
  "/calendar": "calendar",
};

const flowStarts = new Map<string, number>();
const queue: OperationalEvent[] = [];
let flushTimer: number | null = null;
let initialized = false;
let navigationMetricsRecorded = false;

function telemetryAvailable() {
  return import.meta.env.PROD && typeof window !== "undefined" && (window.location.protocol === "https:" || window.location.protocol === "http:");
}

export function initializeTelemetry() {
  if (!telemetryAvailable() || initialized) return;
  initialized = true;
  recordNavigationMetrics();
  window.addEventListener("pagehide", () => flushTelemetry(true));
}

export function trackPage(path: string) {
  const page = PAGE_BY_PATH[normalizePath(path)];
  if (page) enqueue({ name: "page_view", dimension: page });
}

export function trackEvent(name: string) {
  if (SUCCESS_EVENTS.has(name)) enqueue({ name });
}

export function startFlow(name: string) {
  if (!telemetryAvailable() || !FLOWS.has(name)) return;
  flowStarts.set(name, now());
  enqueue({ name: "flow_started", dimension: name });
}

export function finishFlow(name: string, outcome: FlowOutcome) {
  const startedAt = flowStarts.get(name);
  if (startedAt == null || !FLOWS.has(name)) return;
  flowStarts.delete(name);
  enqueue({
    name: "flow_completed",
    dimension: name,
    outcome,
    durationMs: Math.max(0, Math.min(3_600_000, Math.round(now() - startedAt))),
  });
}

export async function trackedFetch(input: RequestInfo | URL, init?: RequestInit): Promise<Response> {
  const startedAt = now();
  const method = normalizeMethod(init?.method ?? (input instanceof Request ? input.method : "GET"));
  const request = requestDescriptor(input, method);
  try {
    const response = await fetch(input, init);
    recordRequest(request.area, method, response.ok ? "success" : "failure", response.status, now() - startedAt);
    if (response.ok && response.status !== 202) recordProductOutcome(request.path, method.toUpperCase());
    return response;
  } catch (error) {
    recordRequest(request.area, method, "network_error", 0, now() - startedAt);
    throw error;
  }
}

function enqueue(event: OperationalEvent) {
  if (!telemetryAvailable()) return;
  queue.push(event);
  if (queue.length >= MAX_BATCH_SIZE) flushTelemetry(false);
  else if (flushTimer == null) flushTimer = window.setTimeout(() => flushTelemetry(false), FLUSH_DELAY_MS);
}

function flushTelemetry(keepalive: boolean) {
  if (flushTimer != null) {
    window.clearTimeout(flushTimer);
    flushTimer = null;
  }
  const events = queue.splice(0, MAX_BATCH_SIZE);
  if (!events.length) return;
  // Deliberately omit cookies and referrer data. The payload contains only
  // allowlisted operational dimensions and is aggregated immediately by the API.
  void fetch(TELEMETRY_ENDPOINT, {
    method: "POST",
    credentials: "omit",
    referrerPolicy: "no-referrer",
    cache: "no-store",
    keepalive,
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ events }),
  }).catch(() => {
    // Telemetry is best-effort and must never interrupt the product.
  });
  if (queue.length && !keepalive) flushTimer = window.setTimeout(() => flushTelemetry(false), 0);
}

function recordNavigationMetrics() {
  if (navigationMetricsRecorded || typeof window === "undefined") return;
  const record = () => {
    if (navigationMetricsRecorded) return;
    const navigation = performance.getEntriesByType("navigation")[0] as PerformanceNavigationTiming | undefined;
    if (!navigation) return;
    navigationMetricsRecorded = true;
    recordMetric("page_load", navigation.duration);
    recordMetric("ttfb", navigation.responseStart - navigation.startTime);
    recordMetric("dom_interactive", navigation.domInteractive - navigation.startTime);
  };
  if (document.readyState === "complete") window.setTimeout(record, 0);
  else window.addEventListener("load", record, { once: true });
}

function recordMetric(metric: string, rawDurationMs: number) {
  enqueue({
    name: "performance_metric",
    dimension: metric,
    durationMs: Math.max(0, Math.min(3_600_000, Math.round(rawDurationMs))),
  });
}

function recordRequest(area: string, method: string, outcome: string, status: number, rawDurationMs: number) {
  enqueue({
    name: "api_request",
    dimension: area,
    method,
    outcome,
    status,
    durationMs: Math.max(0, Math.min(3_600_000, Math.round(rawDurationMs))),
  });
}

function recordProductOutcome(path: string, method: string) {
  if (method === "POST" && /\/auth\/google$/.test(path)) trackEvent("success_account_connected");
  else if (method === "POST" && path.includes("/auth/redeem-premium")) trackEvent("success_pro_trial_activated");
  else if (method === "POST" && path.includes("/credit-checkout/confirm")) trackEvent("success_credits_purchased");
  else if (method === "POST" && path.includes("/support")) trackEvent("success_support_request_sent");
  else if (method !== "GET" && path.includes("/portfolio/watchlist")) trackEvent("success_watchlist_changed");
  else if (method !== "GET" && path.includes("/portfolio/holdings")) trackEvent("success_portfolio_changed");
  else if (method !== "GET" && path.includes("/portfolio/dca-orders")) trackEvent("success_dca_plan_changed");
  else if (method === "POST" && path.includes("/analysis/")) trackEvent("success_ai_analysis_returned");
  else if (method === "POST" && path.includes("/strategy/recommendations")) trackEvent("success_strategy_returned");
  else if (method === "POST" && path.includes("/backtrade/jobs")) trackEvent("success_replay_started");
  else if (method === "PUT" && path.endsWith("/settings")) trackEvent("success_onboarding_completed");
}

function requestDescriptor(input: RequestInfo | URL, method: string) {
  const raw = input instanceof Request ? input.url : String(input);
  try {
    const url = new URL(raw, window.location.origin);
    const path = url.pathname;
    return { path, area: requestArea(path, method) };
  } catch {
    return { path: "unknown", area: `${method}_other` };
  }
}

function requestArea(path: string, method: string) {
  if (path.includes("/buy-timing")) return "buy_timing";
  if (path.includes("/backtrade")) return "replay";
  if (path.includes("/strategy")) return "strategy";
  if (path.includes("/analysis")) return "analysis";
  if (path.includes("/portfolio")) return "portfolio";
  if (path.includes("/auth")) return "auth";
  if (path.includes("/discover") || path.includes("/scanner")) return "discovery";
  if (path.includes("/details")) return "details";
  if (path.includes("/calendar")) return "calendar";
  if (path.includes("/live-trade")) return "live_trade";
  if (path.includes("/bootstrap")) return "bootstrap";
  return `${method}_other`;
}

function normalizeMethod(method: string) {
  const value = method.toLowerCase();
  if (value === "post" || value === "put" || value === "patch" || value === "delete") return value;
  return "get";
}

function normalizePath(path: string) {
  return path.split("?")[0].replace(/\/+$/, "") || "/";
}

function now() {
  return typeof performance !== "undefined" ? performance.now() : Date.now();
}

File: apps/api/routes/telemetry.py
from __future__ import annotations

import hmac
import os
from typing import Any, Literal, Self

from fastapi import APIRouter, Header, HTTPException, Query, Response
from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from internal.store.telemetry import list_operational_telemetry, record_operational_telemetry
from internal.store.db import _is_terminal_libsql_error

router = APIRouter(prefix="/api/telemetry", tags=["operational-telemetry"])

PAGES = {"home", "terms", "privacy", "refunds", "support", "daily_brief", "live_trade", "scanner", "hunt_ai", "calendar"}
FLOWS = {"hunt_signals", "buy_timing", "next_10", "technical_analysis", "intraday_analysis", "strategy_analysis", "analyst_report", "ai_replay"}
AREAS = {"buy_timing", "replay", "strategy", "analysis", "portfolio", "auth", "discovery", "details", "calendar", "live_trade", "bootstrap", "get_other", "post_other", "put_other", "patch_other", "delete_other"}
METRICS = {"page_load", "ttfb", "dom_interactive"}
SUCCESS_EVENTS = {
    "success_dashboard_opened",
    "success_account_connected",
    "success_pro_trial_activated",
    "success_credits_purchased",
    "success_support_request_sent",
    "success_watchlist_changed",
    "success_portfolio_changed",
    "success_dca_plan_changed",
    "success_ai_analysis_returned",
    "success_strategy_returned",
    "success_replay_started",
    "success_onboarding_completed",
}
EVENTS = {"page_view", "flow_started", "flow_completed", "api_request", "performance_metric", *SUCCESS_EVENTS}
DIMENSIONS = {"", *PAGES, *FLOWS, *AREAS, *METRICS}


class TelemetryEvent(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    name: str = Field(min_length=1, max_length=40)
    dimension: str = Field(default="", max_length=40)
    outcome: Literal["", "success", "failure", "abandoned", "network_error"] = ""
    method: Literal["", "get", "post", "put", "patch", "delete"] = ""
    status: int = Field(default=-1, ge=-1, le=599)
    duration_ms: int | None = Field(default=None, alias="durationMs", ge=0, le=3_600_000)

    @field_validator("name")
    @classmethod
    def known_event(cls, value: str) -> str:
        if value not in EVENTS:
            raise ValueError("Unknown operational event")
        return value

    @field_validator("dimension")
    @classmethod
    def known_dimension(cls, value: str) -> str:
        if value not in DIMENSIONS:
            raise ValueError("Unknown operational dimension")
        return value

    @model_validator(mode="after")
    def valid_shape(self) -> Self:
        if self.name == "page_view" and self.dimension not in PAGES:
            raise ValueError("page_view requires a known page")
        if self.name in {"flow_started", "flow_completed"} and self.dimension not in FLOWS:
            raise ValueError("flow event requires a known flow")
        if self.name == "api_request" and self.dimension not in AREAS:
            raise ValueError("api_request requires a known area")
        if self.name == "performance_metric" and self.dimension not in METRICS:
            raise ValueError("performance_metric requires a known metric")
        if self.name in SUCCESS_EVENTS and self.dimension:
            raise ValueError("success events cannot contain a dimension")
        if self.name == "flow_completed" and (not self.outcome or self.duration_ms is None):
            raise ValueError("completed flows require outcome and duration")
        if self.name == "api_request" and (not self.outcome or not self.method or self.duration_ms is None):
            raise ValueError("API requests require outcome, method, and duration")
        return self


class TelemetryBatch(BaseModel):
    model_config = ConfigDict(extra="forbid")

    events: list[TelemetryEvent] = Field(min_length=1, max_length=25)


@router.post("", status_code=204)
def submit_telemetry(payload: TelemetryBatch) -> Response:
    # Deliberately accept no Request object: cookies, account/session identifiers,
    # IP addresses, user agents, and referrers are not read or persisted here.
    try:
        record_operational_telemetry([event.model_dump() for event in payload.events])
    except Exception as exc:
        # Operational telemetry is explicitly best-effort. A stale Turso stream must not
        # create more client retries or compete with account restoration during recovery.
        if not _is_terminal_libsql_error(exc):
            raise
    return Response(status_code=204)


@router.get("/summary")
def telemetry_summary(
    days: int = Query(default=30, ge=1, le=90),
    admin_token: str | None = Header(default=None, alias="X-Telemetry-Admin-Token"),
) -> dict[str, Any]:
    configured_token = os.getenv("TELEMETRY_ADMIN_TOKEN", "").strip()
    if not configured_token:
        raise HTTPException(status_code=404, detail="Not found")
    if not admin_token or not hmac.compare_digest(admin_token, configured_token):
        raise HTTPException(status_code=403, detail="Invalid telemetry admin token")
    return {"days": days, "aggregates": list_operational_telemetry(days)}
